src/pg_copy.c

```diff
@@ -0,0 +1,552 @@
+											 ALLOCSET_DEFAULT_MAXSIZE);
+				
+		/* 
+		 * Construct hash table used for shardId->Connection mapping.
+		 * We can not use connection cache from connection.c used by GteConnection because
+		 * we need to establish multiple connections with each nodes: one connection per shard
+		 */
+		shardToConn = CreateShardToConnectionHash();
+
+		/* init state to read from COPY data source */
+		copyState = BeginCopyFrom(rel, copyStatement->filename,
+								  copyStatement->is_program,
+								  copyStatement->attlist,
+								  copyStatement->options);
+				
+		/* TODO: handle binary mode */
+		if (copyState->binary) { 
+			elog(ERROR, "Copy in binary mode is not currently supported");
+		}
+
+		PG_TRY();
+		{
+			/* Lock all shards in shared mode */
+			shardIntervalList = SortList(shardIntervalList, CompareTasksByShardId);
+			foreach(shardIntervalCell, shardIntervalList)
+			{
+				ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+				ShardId shardId = shardInterval->id;
+				LockShardData(shardId, ShareLock);
+				LockShardDistributionMetadata(shardId, ShareLock);
+			}
+
+			while (nextRowFound)
+			{
+				MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+				nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+				MemoryContextSwitchTo(oldContext);
+					
+				/* write the row to the shard */
+				if (nextRowFound)
+				{	
+					if (columnNulls[partitionColumn->varattno]) 
+					{ 
+						ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+										errmsg("cannot copy row with NULL value "
+											   "in partition column")));
+					} 
+					rightConst->constvalue = columnValues[partitionColumn->varattno];
+					prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+					
+					foreach(shardIntervalCell, prunedList)
+					{
+						ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+						ShardId shardId = shardInterval->id;
+						bool found = false;
+						
+						shardConn = (ShardConnections*)hash_search(shardToConn, &shardInterval->id, HASH_ENTER, &found);
+						if (!found) { 
+							List *finalizedPlacementList = NULL;
+                            int nPlacements = 0;
+
+							finalizedPlacementList = LoadFinalizedShardPlacementList(shardId); 
+                            nPlacements = list_length(finalizedPlacementList);
+                            shardConn->conn = (PGconn**)palloc0(sizeof(PGconn*)*nPlacements);
+							shardConn->nReplicas = nPlacements;
+							shardConn->status = (bool*)palloc0(nPlacements*sizeof(bool));
+                            nPlacements = 0;
+
+							foreach(taskPlacementCell, finalizedPlacementList)
+							{
+								ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+								char *nodeName = taskPlacement->nodeName;
+								char nodePort[16];
+								PGconn* conn = NULL;
+								sprintf(nodePort, "%d", taskPlacement->nodePort);	 
```

Comment:
Hmz, I guess the ConnectToNode API kind of sucks for requiring you to use a string for the port number. Would you mind adding a small wrapper function that does the conversion instead of exposing ConnectToNode directly?